### PR TITLE
ledger snapshots

### PR DESCRIPTION
**Pull request by evanmcc**
Our current approach to following the chain is to download all blocks and then apply them to the ledger.  This takes approximately forever on the current v1 hardware.  snapshots are meant to provide a more concise starting point for hotspots so that they can come online more quickly.  for full nodes, the route of downloading and validating all of the blocks will still be available.

### CODE REVIEW

File: src/blockchain.erl [hunk 1]
```diff
@@ -395,18 +406,14 @@ sync_height(Blockchain = #blockchain{db=DB, temp_blocks=TempBlocksCF, default=De
             end
     end.
 
-%%--------------------------------------------------------------------
-%% @doc
-%% @end
-%%--------------------------------------------------------------------
+-spec ledger() -> blockchain_ledger_v1:ledger().
```
Comment (Vagabond) [inline src/blockchain.erl:409]: hah, finally

File: src/blockchain.erl [hunk 2]
```diff
@@ -610,6 +616,7 @@ add_block(Block, #blockchain{db=DB, blocks=BlocksCF, heights=HeightsCF, default=
                     {ok, NextBlock} ->
                         case blockchain_block:prev_hash(NextBlock) of
                             Hash ->
+                                {ok, Batch} = rocksdb:batch(),
```
Comment (Vagabond) [inline src/blockchain.erl:619]: Maybe we don't need a batch here then?

File: src/blockchain.erl [hunk 3]
```diff
@@ -747,6 +754,47 @@ add_block_(Block, Blockchain, Syncing) ->
                               true -> unvalidated_absorb_and_commit;
                               _ -> absorb_and_commit
                           end,
+                    Ledger = blockchain:ledger(Blockchain),
+                    case blockchain_block_v1:snapshot_hash(Block) of
+                        <<>> ->
+                            ok;
+                        ConsensusHash ->
```
Comment (Vagabond) [inline src/blockchain.erl:761]: Can we break this into a function?

File: src/blockchain.erl [hunk 4]
```diff
@@ -747,6 +754,47 @@ add_block_(Block, Blockchain, Syncing) ->
...
+                                         || N <- lists:seq(DHeight, LedgerHeight)],
+                                        case blockchain_ledger_snapshot_v1:snapshot(Ledger, Blocks) of
+                                            {ok, Snap} ->
+                                                case blockchain_ledger_snapshot_v1:hash(Snap) of
+                                                    ConsensusHash ->
+                                                        ok = blockchain:add_snapshot(Snap, Blockchain);
+                                                    OtherHash ->
+                                                        lager:info("bad snapshot hash: ~p good ~p",
+                                                                   [OtherHash, ConsensusHash]),
+                                                        blockchain_worker:async_reset(Height)
```
Comment (Vagabond) [inline src/blockchain.erl:786]: *casually wields giant hammer*

Maybe make the log message a bit more explicit that we're doing something drastic?

Comment (Vagabond) [reply src/blockchain.erl:786]: I'm not even sure this is good behavior to do automatically, maybe just halting is better?

File: src/blockchain.erl [hunk 5]
```diff
@@ -1373,6 +1443,64 @@ missing_block(#blockchain{db=DB, default=DefaultCF}) ->
             not_found
     end.
 
+-spec add_snapshot(blockchain_ledger_snapshot:snapshot(), blockchain()) ->
+                          ok | {error, any()}.
+add_snapshot(Snapshot, #blockchain{db=DB, snapshots=SnapshotsCF}) ->
```
Comment (Vagabond) [inline src/blockchain.erl:1448]: Should we GC old snapshots, especially on non-full nodes we probably shouldn't try to catch em all?

File: src/blockchain.erl [hunk 6]
```diff
@@ -1394,15 +1522,39 @@ clean(Dir) when is_list(Dir) ->
...
-            Ledger = blockchain_ledger_v1:new(Dir),
+        {ok, DB, [DefaultCF, BlocksCF, HeightsCF, TempBlocksCF, PlausibleBlocksCF, SnapshotCF]} ->
+            Ledger =
+                case Mode of
+                    blessed_snapshot ->
+                        L = blockchain_ledger_v1:new(Dir),
+                        case blockchain_ledger_v1:current_height(L) of
+                            {ok, 1} ->
+                                L;
+                            {ok, CHt} when CHt < 100000000 ->
```
Comment (Vagabond) [inline src/blockchain.erl:1541]: What?

File: src/cli/blockchain_cli_snapshot.erl [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+snapshot_load(["snapshot", "load", Filename], [], []) ->
+    Ret = snapshot_load(Filename),
+    [clique_status:text(io_lib:format("~p", [Ret]))];
+snapshot_load(_, _, _) ->
+    usage.
+
+snapshot_load(Filename) ->
+    {ok, BinSnap} = file:read_file(Filename),
+    %% no reason to hang this forever?
+    %% ok = blockchain_lock:acquire(15000),
```
Comment (Vagabond) [inline src/cli/blockchain_cli_snapshot.erl:113]: Why not timeout?

File: src/ledger/v1/blockchain_ledger_snapshot_v1.erl [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+
+serialize(Snapshot, BlocksP) ->
+    Snapshot1 = case BlocksP of
+                    blocks -> Snapshot;
+                    noblocks -> Snapshot#blockchain_snapshot_v1{blocks = []}
+                end,
+    Bin = term_to_binary(Snapshot1, [{compressed, 9}]),
+    BinSz = byte_size(Bin),
+
+    %% SHA = crypto:hash(sha256, Bin),
```
Comment (Vagabond) [inline src/ledger/v1/blockchain_ledger_snapshot_v1.erl:161]: Why are the hashes commented out?

File: src/cli/blockchain_cli_snapshot.erl (hunk 7 above)
Comment (evanmcc) [reply src/cli/blockchain_cli_snapshot.erl:113]: I wrote this pre-rebase to something that had the timeout and forgot to switch back to it.

File: src/ledger/v1/blockchain_ledger_snapshot_v1.erl (hunk 8 above)
Comment (evanmcc) [reply src/ledger/v1/blockchain_ledger_snapshot_v1.erl:161]: I switched from hashes internal to the framing to hashes external to the framing, more like txns.  I'll remove these leftovers.

File: src/blockchain.erl (hunk 4 above)
Comment (evanmcc) [reply src/blockchain.erl:786]: OK.  I already have halting behavior elsewhere, but perhaps we need a better way to signal that something else is wrong?  Can we affect the light at all anymore?

File: src/blockchain.erl (hunk 5 above)
Comment (evanmcc) [reply src/blockchain.erl:1448]: Yes.  We need a GC for this and for blocks.  I figured we could miss it out for v1 (we'll probably want to set it to something like a month or two anyway).

File: src/blockchain.erl (hunk 6 above)
Comment (evanmcc) [reply src/blockchain.erl:1541]: Hahah ooops.  I did that so this would compile, and then forgot to do the right actual thing.

File: src/blockchain.erl (hunk 2 above)
Comment (evanmcc) [reply src/blockchain.erl:619]: I'm confused.   This is old behavior and the block looks required to keep the various column families in sync?

File: src/blockchain.erl (hunk 3 above)
Comment (evanmcc) [reply src/blockchain.erl:761]: which part?

File: src/blockchain_worker.erl [hunk 9]
```diff
@@ -264,11 +290,12 @@ signed_metadata_fun() ->
 %% gen_server Function Definitions
 %% ------------------------------------------------------------------
 init(Args) ->
-    erlang:process_flag(trap_exit, true),
+    %% erlang:process_flag(trap_exit, true),
```
Comment (Vagabond) [inline src/blockchain_worker.erl]: Uncomment or remove?

File: src/blockchain.erl [hunk 10]
```diff
@@ -774,18 +822,24 @@ add_block_(Block, Blockchain, Syncing) ->
...
-                                  end;
-                           (_, Error) -> Error
-                          end, ok, lists:seq(LedgerHeight+1, BlockchainHeight))
+            replay_blocks(Blockchain, Syncing, LedgerHeight, BlockchainHeight)
     end.
 
+replay_blocks(Chain, Syncing, LedgerHeight, ChainHeight) ->
+    lists:foldl(
+      fun(H, ok) ->
+              {ok, B} = get_block(H, Chain),
```
Comment (vihu) [inline src/blockchain.erl:831]: `get_block` may give you an `{error, _}`

File: src/blockchain.erl [hunk 11]
```diff
@@ -1373,6 +1443,64 @@ missing_block(#blockchain{db=DB, default=DefaultCF}) ->
...
+    case rocksdb:get(DB, SnapshotsCF, <<Height:64/integer-unsigned-big>>, []) of
+       {ok, Hash} ->
+           ?MODULE:get_snapshot(Hash, Blockchain);
+        not_found ->
+            {error, not_found};
+        Error ->
+            Error
+    end.
+
+find_last_snapshot(Blockchain) ->
```
Comment (vihu) [inline src/blockchain.erl:1489]: Maybe add a spec to this fun since it's exported?

File: src/blockchain_swarm.erl [hunk 12]
```diff
@@ -90,7 +90,6 @@ gossip_peers() ->
 %% gen_server Function Definitions
 %% ------------------------------------------------------------------
 init(Args) ->
-    erlang:process_flag(trap_exit, true),
```
Comment (vihu) [inline src/blockchain_swarm.erl:93]: Why was this removed?

File: src/blockchain_worker.erl [hunk 13]
```diff
@@ -70,11 +78,13 @@
...
          sync_pid :: undefined | pid(),
          sync_paused = false :: boolean(),
+         snapshot_info :: undefined | {binary(), integer()},
          gossip_ref = make_ref() :: reference(),
          absorb_info :: undefined | {pid(), reference()},
          absorb_retries = 3 :: pos_integer(),
          resync_info :: undefined | {pid(), reference()},
-         resync_retries = 3 :: pos_integer()
+         resync_retries = 3 :: pos_integer(),
+         mode = startup :: assumed_valid | snapshot | normal | reset
```
Comment (vihu) [inline src/blockchain_worker.erl:87]: `startup` should also be included in the type spec?

File: src/blockchain_worker.erl [hunk 14]
```diff
@@ -504,11 +591,14 @@ code_change(_OldVsn, State, _Extra) ->
     {ok, State}.
 
 terminate(_Reason, #state{blockchain=undefined}) ->
+    ct:pal("blockchain terminate ~p", [_Reason]),
```
Comment (vihu) [inline src/blockchain_worker.erl:594]: Change these and below to `lager:warning` or something before landing?

File: src/blockchain_worker.erl [hunk 15]
```diff
@@ -517,12 +607,31 @@ terminate(_Reason, #state{blockchain=Chain}) ->
...
+    maybe_sync_blocks(State);
+maybe_sync(#state{mode = snapshot, blockchain = Chain, sync_pid = Pid} = State) ->
+    Ledger = blockchain:ledger(Chain),
+    case blockchain_ledger_v1:current_height(Ledger) of
+        %% still waiting for the background process to download and
+        %% install the ledger snapshot
+        {ok, 1} when Pid /= undefined ->
+            reset_sync_timer(State);
+        {ok, 1} ->
+            {Hash, Height} = State#state.snapshot_info,
```
Comment (vihu) [inline src/blockchain_worker.erl:620]: `State#state.snapshot_info` may return `undefined`. Maybe pattern match it out in the fun head if so and handle accordingly?

Comment (vihu) [reply src/blockchain_worker.erl:620]: Maybe better to have getter and setters for `snapshot_info` and other newly introduced state fields?

File: src/blockchain_worker.erl [hunk 16]
```diff
@@ -674,8 +844,42 @@ get_assume_valid_height_and_hash() ->
     end.
 
 load_chain(SwarmTID, BaseDir, GenDir) ->
-    AssumedValidBlockHashAndHeight = get_assume_valid_height_and_hash(),
-    case blockchain:new(BaseDir, GenDir, AssumedValidBlockHashAndHeight) of
+    QuickSyncMode = application:get_env(blockchain, quick_sync_mode, assumed_valid),
+    QuickSyncData =
+        case QuickSyncMode of
+            assumed_valid ->
```
Comment (vihu) [inline src/blockchain_worker.erl:850]: Split the cases below into separate functions maybe?

File: src/blockchain_swarm.erl (hunk 12 above)
Comment (evanmcc) [reply src/blockchain_swarm.erl:93]: it was hiding some crash issues I was seeing during development, so I got rid of it.  I think that it's a hindrance to debugging, and should only be used on supervising processes.

File: src/blockchain_worker.erl (hunk 13 above)
Comment (evanmcc) [reply src/blockchain_worker.erl:87]: no, I'll clean that up.

File: src/blockchain.erl (hunk 4 above)
Comment (evanmcc) [reply src/blockchain.erl:786]: making it so we halt.

File: src/blockchain_worker.erl (hunk 14 above)
Comment (evanmcc) [reply src/blockchain_worker.erl:594]: oops, fixed.

File: src/blockchain_worker.erl (hunk 15 above)
Comment (evanmcc) [reply src/blockchain_worker.erl:620]: I don't really like getters and setters for internal use.   

Rather than match it out, I double checked that everywhere we go into `snapshot` we populate the `snapshot_info`.  init wasn't doing it correctly, that's fixed.

File: src/blockchain.erl (hunk 10 above)
Comment (evanmcc) [reply src/blockchain.erl:831]: since this code path won't get hit except on manual intervention, that seems OK?

File: src/blockchain.erl (hunk 2 above)
Comment (Vagabond) [reply src/blockchain.erl:619]: Oh, I thought we were immediately committing the batch, sorry.

File: src/blockchain.erl (hunk 3 above)
Comment (Vagabond) [reply src/blockchain.erl:761]: the handling if we see a snapshot hash? just so this function doesn't get too deep?

File: src/blockchain.erl [hunk 17]
```diff
@@ -1373,6 +1443,65 @@ missing_block(#blockchain{db=DB, default=DefaultCF}) ->
...
+    case rocksdb:get(DB, SnapshotsCF, <<Height:64/integer-unsigned-big>>, []) of
+       {ok, Hash} ->
+           ?MODULE:get_snapshot(Hash, Blockchain);
+        not_found ->
+            {error, not_found};
+        Error ->
+            Error
+    end.
+
+-spec find_last_snapshot(blockchain()) -> undefined | tuple() | {error, any()}.
```
Comment (Vagabond) [inline src/blockchain.erl:1489]:
```suggestion
-spec find_last_snapshot(blockchain()) -> undefined | {non_neg_integer(), blockchain_block:hash(), binary()} | {error, any()}.
```

File: src/blockchain.erl [hunk 18]
```diff
@@ -774,18 +789,63 @@ add_block_(Block, Blockchain, Syncing) ->
...
-                                          run_absorb_block_hooks(Syncing, H, Blockchain);
-                                      {error, Reason} ->
-                                          lager:error("Error absorbing transaction, Ignoring Hash: ~p, Reason: ~p", [blockchain_block:hash_block(Block), Reason])
-                                  end;
-                           (_, Error) -> Error
-                          end, ok, lists:seq(LedgerHeight+1, BlockchainHeight))
+            replay_blocks(Blockchain, Syncing, LedgerHeight, BlockchainHeight)
+    end.
+
+check_snapshot(ConsensusHash, MyAddress, Signers,
```
Comment (Vagabond) [inline src/blockchain.erl:795]: check is sort of a misnomer here? we're generating a local copy of a snapshot from the chain?

File: src/blockchain_worker.erl [hunk 19]
```diff
@@ -313,6 +355,38 @@ handle_call({new_ledger, Dir}, _From, State) ->
     Ledger1 = blockchain_ledger_v1:new(Dir),
     {reply, {ok, Ledger1}, State};
 
+handle_call({install_snapshot, Hash, Snapshot}, _From,
+            #state{blockchain = Chain, mode = Mode, swarm = Swarm} = State) ->
+    %% I don't think that we want to auto-repair right now, do default
+    %% this to disabled.
+    Halt = application:get_env(blockchain, halt_on_reset, true),
+    case Mode == reset andalso Halt of
```
Comment (Vagabond) [inline src/blockchain_worker.erl:363]: I don't have a good feel for what 'halt on reset' means here.

Also, do we ever set mode to be reset anywhere?

Comment (evanmcc) [reply src/blockchain_worker.erl:363]: well, this code currently stops the worker, but I'm gonna disable it shortly.  At this point all it will do is log.  Eventually it should either auto-repair or stop sync and go idle.
